[content-visibility] Lazily layout children

rwlbuis opened the pull request:
#### faa0935aebd5fb9dfde94bf8c91262fd70479bc4
<pre>
[content-visibility] Lazily layout children
<a href="https://bugs.webkit.org/show_bug.cgi?id=237369">https://bugs.webkit.org/show_bug.cgi?id=237369</a>

Reviewed by Tim Nguyen.

This patch supports skipping layout of children in hidden content-visibility subtrees
for normal layout, since they do not count as content [1].

However DOM operations that require measuring do need an accurate layout, so
introduce the helper class ContentVisibilityForceLayoutScope to override (&quot;forced layout&quot;)
the skipped layout mechanism for hidden content-visibility subtrees.

Roughly the skipping layout of children falls into two catagories:
- general block flows (note flexbox is part of this)
- grid containers

The existing perf tests under ContentVisibility/ improve a lot with this patch (up to 100x), but
the scrolling variants have only slight improvement since scrolling content into the viewport
means falling back to full layout.
Improving the perf tests of grid based content-visibility perf tests will be addressed in a future PR.

[1] <a href="https://w3c.github.io/csswg-drafts/css-contain/#skips-its-contents">https://w3c.github.io/csswg-drafts/css-contain/#skips-its-contents</a>

* LayoutTests/TestExpectations:
* Source/WebCore/css/ComputedStyleExtractor.cpp: make sure layout is up-to-date for hidden content-visibility subtree.
(WebCore::ComputedStyleExtractor::propertyValue const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout): forces layout on hidden content-visibility subtree if needed.
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::scrollLeft):
(WebCore::Element::scrollTop):
(WebCore::Element::setScrollLeft):
(WebCore::Element::setScrollTop):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingClientRect):
(WebCore::Element::innerText):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::width):
(WebCore::HTMLImageElement::height):
(WebCore::HTMLImageElement::x const):
(WebCore::HTMLImageElement::y const):
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/page/LocalFrameViewLayoutContext.h: keep track of forced layout.
(WebCore::LocalFrameViewLayoutContext::needsSkippedContentLayout const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::layoutPositionedObject): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutInFlowChildren): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::clearNeedsLayoutForDescendants): add helper to clear layout for subtree.
(WebCore::RenderElement::layoutIfNeeded): take isSkippedContentForLayout into account.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::layoutIfNeeded): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutPositionedObject): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderLayoutState.cpp: Add helper to force layout if there is a hidden content-visibility subtree.
(WebCore::ContentVisibilityForceLayoutScope::ContentVisibilityForceLayoutScope):
(WebCore::ContentVisibilityForceLayoutScope::~ContentVisibilityForceLayoutScope):
* Source/WebCore/rendering/RenderLayoutState.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContentForLayout const): same as isSkippedContent but takes into account needsSkippedContentLayout.
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
(WebCore::SVGGeometryElement::isPointInFill):
(WebCore::SVGGeometryElement::isPointInStroke):
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::SVGLocatable::getBBox):
(WebCore::SVGLocatable::computeCTM):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::getBBox):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::getIntersectionList):
(WebCore::SVGSVGElement::getEnclosureList):
(WebCore::SVGSVGElement::checkIntersection):
(WebCore::SVGSVGElement::checkEnclosure):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::getNumberOfChars):
(WebCore::SVGTextContentElement::getComputedTextLength):
(WebCore::SVGTextContentElement::getCharNumAtPosition):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::updateLayoutIgnorePendingStylesheetsAndRunPostLayoutTasks):

Canonical link: <a href="https://commits.webkit.org/268310@main">https://commits.webkit.org/268310@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/784bc3e674082d945826aa7ce4e62e55a882516e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19307 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19727 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18065 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19540 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22996 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19850 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19526 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22066 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16763 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17573 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17825 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21877 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18340 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15528 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17473 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4617 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21833 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18170 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/page/FrameViewLayoutContext.h:
```diff
@@ -82,6 +82,8 @@ class FrameViewLayoutContext {
     bool isInLayout() const { return layoutPhase() != LayoutPhase::OutsideLayout; }
     bool isInRenderTreeLayout() const { return layoutPhase() == LayoutPhase::InRenderTreeLayout; }
     bool inPaintableState() const { return layoutPhase() != LayoutPhase::InRenderTreeLayout && layoutPhase() != LayoutPhase::InViewSizeAdjust && (layoutPhase() != LayoutPhase::InPostLayout || inAsynchronousTasks()); }
+    bool isLayoutForced() const { return m_layoutForced; }
```
"Force" here is too vague. Forced by what? This needs a name that clearly associates it with skipped content.

rwlbuis replied:
Fair enough, I renamed to something more descriptional.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2282,6 +2282,31 @@ bool Document::updateStyleIfNeeded()
+{
+    ContentVisibilityForceLayoutScope scope(element);
+    updateLayoutIgnorePendingStylesheets();
+}
```
Can we not introduce this new method and instead have a single entry layout entry point that takes an OptionSet ?

It's possible someone might want to update layout not ignoring pending stylesheets, while also taking in account content-visibility. Also, with a single point of entry with options, people will have to think about the `OptionSet<>` and whether they want content-visibility to be taken in account every time they add a new layout.

rwlbuis replied:
It is a good idea in principle I think. I implemented it, but since the Element also needs to be passed, I am not sure if it is needed in the end, since we can have a bit strange situations where ContentVisiblityForceLayout flag can be passed but not Element, or vice-versa.

However integrating the ContentVisibilityForceLayoutScope helper functionality deeper (right now it is in updateLayout) seems like a good move, ContentVisibilityForceLayoutScope is much smaller now and maybe we do not even need it in the end since it is only used in one place.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2282,6 +2282,31 @@ bool Document::updateStyleIfNeeded()
+        if (context.renderer() && context.renderer()->style().skippedContentReason().has_value()) {
+            context.renderer()->setNeedsLayout();
+            m_document = context.document();
+            m_document->view()->layoutContext().forceSkippedContentLayout(true);
```
The name `forceSkippedContentLayout` is slightly confusing, we're not forcing a layout, but setting a flag.

Maybe `setNeedsSkippedContentLayout()` && `needsSkippedContentLayout()`

rwlbuis replied:
Yes, sounds more accurate, will have a stab at that tomorrow.

nt1m commented on Source/WebCore/rendering/RenderBlockFlow.cpp:
```diff
@@ -746,12 +746,15 @@ void TextBoxTrimmer::adjustTextBoxTrimAfterLayout()
 void RenderBlockFlow::layoutInFlowChildren(bool relayoutChildren, LayoutUnit& repaintLogicalTop, LayoutUnit& repaintLogicalBottom, LayoutUnit& maxFloatLogicalBottom)
 {
-    if (!firstChild()) {
+    if (!firstChild() || (firstChild()->isSkippedContentForLayout() && !normalChildNeedsLayout())) {
```
Same question here: can you elaborate why these changes are needed? Are the changes in `layoutIfNeeded` not good enough to skip work here?

rwlbuis replied:
This change is a bit vague indeed, it improves performance but I will check tomorrow if there is a more logical place that gives the same performance.

nt1m commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
The changes in this file seem a bit arbitrary, can you elaborate why they are needed? Are the changes in `layoutIfNeeded` not good enough to skip work here?

rwlbuis replied:
These are needed to skip unneeded work in the grid algorithms and make a lot of difference in the grid perf tests.

nt1m commented on Source/WebCore/page/LocalFrameViewLayoutContext.h:
```diff
@@ -93,6 +93,9 @@ class LocalFrameViewLayoutContext {
 
+    bool isSkippedContentLayoutForced() const { return m_isSkippedContentLayoutForced; }
+    void forceSkippedContentLayout(bool f) { m_isSkippedContentLayoutForced = f; }
+
```
These should probably be on `RenderLayoutState`

rwlbuis replied:
I did not see how that would work but have an idea now, will try to implement it tomorrow.

nt1m commented on Source/WebCore/svg/SVGGeometryElement.cpp:
```diff
@@ -47,7 +47,7 @@ SVGGeometryElement::SVGGeometryElement(const QualifiedName& tagName, Document& d
 float SVGGeometryElement::getTotalLength() const
 {
-    document().updateLayoutIgnorePendingStylesheets();
+    document().updateLayoutForNode(*this);
```
Does content-visibility properly work on SVG in general?

rwlbuis replied:
I have to ask around, since officially svg does not generate boxes, maybe not.

rwlbuis replied:
Playing around with this in chromium, it seems supported. I'll try to check whether this is by design, but I do think there is no proper test for this, since all tests seem to use outer svg or its container for setting content-visibility, not the svg shapes/text/groups etc.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -3794,7 +3799,7 @@ ExceptionOr<void> Element::setInnerHTML(const String& html)
 {
     // We need to update layout, since plainText uses line boxes in the render tree.
-    document().updateLayoutIgnorePendingStylesheets();
+    document().updateLayoutForNode(*this);
```
Do we need this for innerText() ? Since skipped content is skipped from innerText, I'm not sure we need to lay it out (I may be wrong though)

rwlbuis replied:
Indeed it is not needed, the important thing is to make sure updateRelevancyOfContentVisibilityElements is called and style is up-to-date, fixed.

nt1m commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1318,6 +1318,8 @@ static HashSet<TreeScope*> collectAncestorTreeScopeAsHashSet(Node& node)
 
 int Element::offsetLeftForBindings()
```
Is it correct to do this just for script? Do we know if some internal callers may want skipped content layout as well ?

rwlbuis replied:
It looks like this is only used internally bu spatial navigation code. Anyway it should be safe to move this code to offsetLeft method, done.

nt1m reviewed (commented):
Here's some high-level feedback after discussing with other folks:
- We'd like to make sure future code also takes content-visibility in account, potentially through a common layout entry point that takes an OptionSet that needs to be thought out.

- Do we know of places that are not script that might want skipped content to be laid out? (find in page for instance)

smfr commented on Source/WebCore/html/ImageInputType.cpp:
```diff
@@ -175,6 +175,8 @@ unsigned ImageInputType::height() const
     ASSERT(element());
     Ref<HTMLInputElement> element(*this->element());
 
+    ContentVisibilityForceLayoutScope scope(element.get());
```
These seems to be sprinkled around fairly arbitrarily. How will future patches know when this is needed? We need to factor the code in such a way that it's hard to make mistakes in future.

rwlbuis replied:
For this case I was not sure ContentVisibilityForceLayoutScope made sense in updateLayout, hence I kept it out, but looking at updateLayout again, it does make sense I think to use ContentVisibilityForceLayoutScope in updateLayout (and only there), so I was able to remove all other ContentVisibilityForceLayoutScope uses.

nt1m commented on Source/WebCore/rendering/RenderElement.cpp:
```diff
@@ -2288,5 +2288,22 @@ bool RenderElement::hasEligibleContainmentForSizeQuery() const
     return false;
 }
 
+void RenderElement::clearNeedsLayoutForDescendents()
```
nit: typo

```suggestion
void RenderElement::clearNeedsLayoutForDescendants()
```

rwlbuis replied:
Good eyes, fixed.

nt1m commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
Can we split out these changes to a new patch?

rwlbuis replied:
Sure, I filed https://bugs.webkit.org/show_bug.cgi?id=261825 to track it and will create a PR for that after this PR lands.

nt1m commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -2304,13 +2305,13 @@ bool Document::updateStyleIfNeeded()
 {
     layoutOptions.add(LayoutOptions::IgnorePendingStylesheets);
-    updateLayout(layoutOptions);
+    updateLayout(layoutOptions, layoutOptions.contains(LayoutOptions::ContentVisibilityForceLayout) ? context : nullptr);
```
Passing the context unconditionally here probably doesn't hurt. It could be used for different optimizations unrelated to content visibility

rwlbuis replied:
You are right, cleaner that way, adjusted.

rwlbuis committed:
[content-visibility] Lazily layout children
https://bugs.webkit.org/show_bug.cgi?id=237369

Reviewed by Tim Nguyen.

This patch supports skipping layout of children in hidden content-visibility subtrees
for normal layout, since they do not count as content [1].

However DOM operations that require measuring do need an accurate layout, so
introduce the helper class ContentVisibilityForceLayoutScope to override ("forced layout")
the skipped layout mechanism for hidden content-visibility subtrees.

Roughly the skipping layout of children falls into two catagories:
- general block flows (note flexbox is part of this)
- grid containers

The existing perf tests under ContentVisibility/ improve a lot with this patch (up to 100x), but
the scrolling variants have only slight improvement since scrolling content into the viewport
means falling back to full layout.
Improving the perf tests of grid based content-visibility perf tests will be addressed in a future PR.

[1] https://w3c.github.io/csswg-drafts/css-contain/#skips-its-contents

* LayoutTests/TestExpectations:
* Source/WebCore/css/ComputedStyleExtractor.cpp: make sure layout is up-to-date for hidden content-visibility subtree.
(WebCore::ComputedStyleExtractor::propertyValue const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout): forces layout on hidden content-visibility subtree if needed.
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::scrollLeft):
(WebCore::Element::scrollTop):
(WebCore::Element::setScrollLeft):
(WebCore::Element::setScrollTop):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingClientRect):
(WebCore::Element::innerText):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::width):
(WebCore::HTMLImageElement::height):
(WebCore::HTMLImageElement::x const):
(WebCore::HTMLImageElement::y const):
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/page/LocalFrameViewLayoutContext.h: keep track of forced layout.
(WebCore::LocalFrameViewLayoutContext::needsSkippedContentLayout const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::layoutPositionedObject): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutInFlowChildren): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::clearNeedsLayoutForDescendants): add helper to clear layout for subtree.
(WebCore::RenderElement::layoutIfNeeded): take isSkippedContentForLayout into account.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::layoutIfNeeded): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutPositionedObject): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderLayoutState.cpp: Add helper to force layout if there is a hidden content-visibility subtree.
(WebCore::ContentVisibilityForceLayoutScope::ContentVisibilityForceLayoutScope):
(WebCore::ContentVisibilityForceLayoutScope::~ContentVisibilityForceLayoutScope):
* Source/WebCore/rendering/RenderLayoutState.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContentForLayout const): same as isSkippedContent but takes into account needsSkippedContentLayout.
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
(WebCore::SVGGeometryElement::isPointInFill):
(WebCore::SVGGeometryElement::isPointInStroke):
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::SVGLocatable::getBBox):
(WebCore::SVGLocatable::computeCTM):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::getBBox):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::getIntersectionList):
(WebCore::SVGSVGElement::getEnclosureList):
(WebCore::SVGSVGElement::checkIntersection):
(WebCore::SVGSVGElement::checkEnclosure):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::getNumberOfChars):
(WebCore::SVGTextContentElement::getComputedTextLength):
(WebCore::SVGTextContentElement::getCharNumAtPosition):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::updateLayoutIgnorePendingStylesheetsAndRunPostLayoutTasks):

Canonical link: https://commits.webkit.org/268310@main

nt1m commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -953,6 +953,12 @@ LayoutUnit RenderBlock::marginIntrinsicLogicalWidthForChild(RenderBox& child) co
 
 void RenderBlock::layoutPositionedObject(RenderBox& r, bool relayoutChildren, bool fixedPositionObjectsOnly)
 {
+    if (isSkippedContentRoot()) {
```
I should have asked this earlier, but is this correct when `ContentVisibilityForceLayout` is on ? Or is this unreachable/irrelevant code when `ContentVisibilityForceLayout` is on?
